### PR TITLE
feat: add support for BigQuery DataFrames. Set `context.engine` to 'bigframes' to support query results larger than 10 GB

### PR DESCRIPTION
**Pull request by sycai**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -28,6 +28,7 @@
 # 'Development Status :: 5 - Production/Stable'``
 release_status = "Development Status :: 4 - Beta"
 dependencies = [
+    "bigframes >= 1.17.0",
```
Comment (tswast) [inline setup.py:31]: Please don't.

In the design I shared that this should be an optional dependency. That means it should be an "extra".

Comment (sycai) [reply setup.py]: Ack. Moved it to extra.

File: owlbot.py [hunk 2]
```diff
@@ -56,6 +56,7 @@
         # Multi-processing note isn't relevant, as bigquery-magics is responsible for
         # creating clients, not the end user.
         "docs/multiprocessing.rst",
+        "noxfile.py",
```
Comment (tswast) [inline owlbot.py:59]: Please don't modify the noxfile in this way. There is a feature in the template that does what we want.

https://github.com/googleapis/synthtool/blob/master/synthtool/gcp/templates/python_library/noxfile.py.j2#L75C1-L75C27

Use unit_test_extras_by_python to make sure we have some Python versions where we test without extras and some where we test with.

File: noxfile.py [hunk 3]
```diff
@@ -217,6 +224,9 @@ def unit(session, protobuf_implementation):
     if protobuf_implementation == "cpp":
         session.install("protobuf<4")
 
+    if install_bigframes:
+        session.install("bigframes >= 1.17.0")
```
Comment (tswast) [inline noxfile.py:228]: Please use the "extras". Don't install this manually.

File: setup.py [hunk 4]
```diff
@@ -54,6 +54,7 @@
         "grpcio >= 1.47.0, < 2.0dev",
         "grpcio >= 1.49.1, < 2.0dev; python_version>='3.11'",
     ],
+    "bigframes": ["bigframes >= 1.17.0"],
```
Comment (tswast) [inline setup.py:57]: `pip` will always install the latest version of bigframes. Let's make sure we test our compatibility with the minimum supported bigframes by adding `bigframes==1.17.0` to https://github.com/googleapis/python-bigquery-magics/blob/main/testing/constraints-3.9.txt

Comment (sycai) [reply setup.py:57]: Done

File: owlbot.py (hunk 2 above)
Comment (sycai) [reply owlbot.py]: Change reverted

File: noxfile.py (hunk 3 above)
Comment (sycai) [reply noxfile.py]: Sounds good. Updated owlbot.py instead

File: bigquery_magics/bigquery.py [hunk 5]
```diff
@@ -444,6 +444,45 @@ def _split_args_line(line: str) -> Tuple[str, str]:
...
 
+def _query_with_bigframes(query: str, params: List[Any], args: Any):
+    if args.dry_run:
+        raise ValueError("Dry run is not supported by bigframes engine.")
+
+    if bpd is None:
+        raise ValueError("Bigframes package is not installed.")
+
+    bpd.options.bigquery.project = context.project
+    bpd.options.bigquery.credentials = context.credentials
```
Comment (tswast) [inline bigquery_magics/bigquery.py:455]: I'm worried these might fail if the bigframes session has already started. Maybe that's what we want to happen?

Alternatively, we could explicitly create a bigframes Session object, but then whatever DataFrame we produce won't be compatible with other DataFrames in the notebook.

**Review by tswast — APPROVED** [review]
Thanks!

**Pull request merged**